src/storage-client/src/statistics.rs

```diff
@@ -18,50 +18,537 @@
+        .with_column("snapshot_records_staged", ScalarType::UInt64.nullable(true))
+        //
+        // Non-resetting gauges
+        //
+        // Whether or not the snapshot for the source has been committed. Never resets.
+        .with_column("snapshot_committed", ScalarType::Bool.nullable(false))
+        // The following are not yet reported by sources and have 0 or `NULL` values.
+        // They have been added here to reduce churn changing the schema of this collection.
+        //
+        // A gauge of the number of _values_ (source defined unit) available to be read from upstream.
+        // Never resets. Not to be confused with any of the counters above.
+        .with_column("offset_known", ScalarType::UInt64.nullable(false))
+        // A gauge of the number of _values_ (source defined unit) we have committed.
+        // Never resets. Not to be confused with any of the counters above.
+        .with_column("offset_committed", ScalarType::UInt64.nullable(false))
+});
+
+pub static MZ_SINK_STATISTICS_RAW_DESC: Lazy<RelationDesc> = Lazy::new(|| {
+    RelationDesc::empty()
+        // Id of the sink.
+        .with_column("id", ScalarType::String.nullable(false))
+        //
+        // Counters
+        //
+        // A counter of the messages we have staged to upstream.
+        // Never resets.
+        .with_column("messages_staged", ScalarType::UInt64.nullable(false))
+        // A counter of the messages we have committed.
+        // Never resets.
+        .with_column("messages_committed", ScalarType::UInt64.nullable(false))
+        // A counter of the bytes we have staged to upstream.
+        // Never resets.
+        .with_column("bytes_staged", ScalarType::UInt64.nullable(false))
+        // A counter of the bytes we have committed.
+        // Never resets.
+        .with_column("bytes_committed", ScalarType::UInt64.nullable(false))
+});
+
+// Types of statistics (counter and various types of gauges), that have different semantics
+// when sources/sinks are reset.
+
+/// A counter that never resets.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct Counter(u64);
+
+impl Counter {
+    /// Merge `self` with another value.
+    /// Used to merge values across workers.
+    pub fn merge(&mut self, other: Self) {
+        // Sum across workers.
+        self.0 += other.0
+    }
+
+    /// Incorporate this value with another.
+    /// Used to update the values written to the builtin tables.
+    pub fn incorporate(&mut self, other: Self, _field_name: &'static str) {
+        // Always add the new value to the existing one.
+        self.0 += other.0
+    }
+}
+
+impl From<u64> for Counter {
+    fn from(f: u64) -> Self {
+        Counter(f)
+    }
+}
+
+/// A trait defining the reset semantics of various types of gauges.
+pub trait GaugeSemantics {
+    /// Merge `self` with another value.
+    /// Used to merge values across workers.
+    fn merge(&mut self, other: Self);
```

Comment:
As you predicted, I have some suggestions on this trait. My suggestion is that instead of always trying to merge measurements into a single value on worker 0 we could have worker 0 initialize a vector of `vec![<the metric>; num_workers]` such that it always holds the full detail from every worker.

Then you can have two functions:

```rust
trait StorageMetric {
    /// Summarizes a set of measurements into a single representative value.
    /// Typically this function is used to summarize the measurements collected from each worker.
    fn summarize(values: &[Self]) -> Self;
    
    /// Incorporate this value with another.
    fn incorporate(&mut self, other: Self);
}
```

This is so that we can correctly record the max metrics which if I understood the PR correctly can currently be lost or misreported

Reply:
hmm, yes thats possible, but requires some more finagling, ill see if its feasible!

Reply:
@petrosagg the max metrics can't be lost/misreported, but the api's as given can easily be misused

I think with something like

```
fn summarize<I: Iterator<&Self>>(values: I) -> Self;
```

your suggestion is possible, and much cleaner, ill try that!